consensus/safety-rules/src/safety_rules.rs

```diff
@@ -308,44 +285,38 @@ impl TSafetyRules for SafetyRules {
         self.verify_qc(proposed_block.quorum_cert())?;
         proposed_block.validate_signature(&self.epoch_state()?.verifier)?;
 
-        self.check_and_update_preferred_round(proposed_block.quorum_cert())?;
-        self.check_last_vote_round(proposed_block.block_data())?;
+        self.verify_and_update_preferred_round(proposed_block.quorum_cert())?;
+        self.verify_last_vote_round(proposed_block.block_data())?;
+
         let vote_data = self.extension_check(vote_proposal)?;
         self.persistent_storage
             .set_last_voted_round(proposed_block.round())?;
+
         let validator_signer = self.signer()?;
         Ok(Vote::new(
             vote_data,
             validator_signer.author(),
             self.construct_ledger_info(proposed_block),
-            &validator_signer,
+            validator_signer,
         ))
     }
 
     fn sign_proposal(&mut self, block_data: BlockData) -> Result<Block, Error> {
         debug!("Incoming proposal to sign.");
+        self.signer()?;
```

Comment:
nit: I'd avoid doing the check twice, e.g.,

let validator_signer = self.signer()?;
...

Ok(Block...
    block_data,
    validator_signer
)

Reply:
Rust, references, and mutability :/

That was how I had it originally until the compiler informed me how great it is :P